components/datetime/src/pattern/item/ule.rs

```diff
@@ -0,0 +1,215 @@
+/// │     │          Unicode Code Point             │ Literal
+/// ├─┬───┴─────────┬───────────────┬───────────────┤
+/// │X│             │  FieldSymbol  │  FieldLength  │ Field
+/// └─┴─────────────┴───────────────┴───────────────┘
+///  ▲
+///  │
+///  Variant Discriminant
+///
+/// # Optimization
+///
+/// This model is optimized for efficient packaging of the `PatternItem` elements
+/// and performant deserialization from the `PatternItemULE` to `PatternItem` type.
+///
+/// # Constraints
+///
+/// The model leaves at most 8 `PatternItem` variants, limits the number of possible
+/// field types and symbols to 16 each and limits the number of length variants to 128.
+///
+/// [`Unicode Code Point`]: http://www.unicode.org/versions/latest/
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct PatternItemULE([u8; 3]);
+
+impl PatternItemULE {
+    #[inline]
+    fn determine_field_from_u8(byte: u8) -> bool {
+        byte & 0b1000_0000 != 0
+    }
+
+    #[inline]
+    fn bytes_in_range(value: (&u8, &u8, &u8)) -> bool {
+        match Self::determine_field_from_u8(*value.0) {
+            true => fields::Field::bytes_in_range(value.1, value.2),
+            false => {
+                let u = u32::from_be_bytes([0x00, *value.0, *value.1, *value.2]);
+                char::try_from(u).is_ok()
+            }
+        }
+    }
+}
+
+unsafe impl ULE for PatternItemULE {
+    type Error = &'static str;
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let mut chunks = bytes.chunks_exact(3);
+
+        if !chunks.all(|c| Self::bytes_in_range((&c[0], &c[1], &c[2])))
+            || !chunks.remainder().is_empty()
+        {
+            return Err("Invalid byte sequence");
+        }
+        Ok(())
+    }
+}
+
+impl AsULE for PatternItem {
+    type ULE = PatternItemULE;
+
+    #[inline]
+    fn as_unaligned(&self) -> Self::ULE {
+        match self {
+            Self::Field(field) => {
+                PatternItemULE([0b1000_0000, field.symbol.idx(), field.length as u8])
+            }
+            Self::Literal(ch) => {
+                let u = *ch as u32;
+                let bytes = u.to_be_bytes();
+                PatternItemULE([bytes[1], bytes[2], bytes[3]])
+            }
+        }
+    }
+
+    #[inline]
+    fn from_unaligned(unaligned: &Self::ULE) -> Self {
+        let value = unaligned.0;
+        match PatternItemULE::determine_field_from_u8(value[0]) {
+            false => {
+                let u = u32::from_be_bytes([0x00, value[0], value[1], value[2]]);
+                PatternItem::Literal(char::try_from(u).unwrap())
+            }
+            true => {
+                let symbol = fields::FieldSymbol::from_idx(value[1]).unwrap();
+                let length = fields::FieldLength::from_idx(value[2]).unwrap();
+                let field = fields::Field { symbol, length };
+                PatternItem::Field(field)
+            }
+        }
+    }
+}
+
+unsafe impl EqULE for PatternItem {}
```

Comment:
Issue: `EqULE` is not safe to implement, because PatternItem does not have the same byte layout as PatternItemULE.

Note that `EqULE` does not really do anything except unlock the `ZeroVec::try_from_slice` function (you can still use `ZeroVec::clone_from_slice` without `EqULE`).